scripts/init-db/gitlab_db_init.py

```diff
@@ -0,0 +1,105 @@
+        self.gitlab_url = self.gitlab_url.rstrip("/")
+
+    @classmethod
+    def from_env(cls):
+        return cls(
+            db_host=os.environ["DB_HOST"],
+            db_admin_username=os.environ["DB_ADMIN_USERNAME"],
+            db_admin_password=os.environ["DB_ADMIN_PASSWORD"],
+            db_port=int(os.environ.get("DB_PORT", 5432)),
+            gitlab_oauth_client_secret=os.environ["GITLAB_OAUTH_CLIENT_SECRET"],
+            gitlab_oauth_client_id=os.environ["GITLAB_OAUTH_CLIENT_ID"],
+            gitlab_db_password=os.environ["GITLAB_DB_PASSWORD"],
+            gitlab_db_username=os.environ["GITLAB_DB_USERNAME"],
+            gitlab_db_name=os.environ["GITLAB_DB_NAME"],
+            gitlab_url=os.environ["GITLAB_URL"],
+            renku_url=os.environ["RENKU_URL"],
+        )
+
+
+def main():
+    config = Config.from_env()
+    # Wait for services to come online
+    postgres_db_connection = get_db_connection(
+        user=config.db_admin_username,
+        password=config.db_admin_password,
+        host=config.db_host,
+        port=config.db_port,
+    )
+    postgres_db_connection.set_session(autocommit=True)
+
+    # Run queries
+    logging.info("Creating the gitlab postgres database...")
+    db_init = DatabaseInit(
+        config.gitlab_db_username,
+        config.gitlab_db_name,
+        config.gitlab_db_password,
+        postgres_db_connection,
+        ["pg_trgm", "btree_gist"],
+        config.db_admin_username,
+    )
+    db_init.create_database()
+    gitlab_db_connection = get_db_connection(
+        user=config.db_admin_username,
+        password=config.db_admin_password,
+        host=config.db_host,
+        port=config.db_port,
+        database=config.gitlab_db_name,
+    )
+    # NOTE: Database extensions do not get created in transactions
+    gitlab_db_connection.set_session(autocommit=True)
+    db_init.set_connection(gitlab_db_connection)
+    db_init.set_extensions_and_roles()
+    gitlab_callback_uris = " ".join(
+        [
+            f"{config.renku_url}/login/redirect/gitlab",
+            f"{config.renku_url}/api/auth/gitlab/token",
+        ]
+    )
+    # NOTE: Querying for the base url switches the request to port 443 and the service does not
+    # expose port 443.
+    gitlab_is_online(config.gitlab_url + "/help")
+    with gitlab_db_connection.cursor() as curs:
+        logging.info(f"Cleaning up client id {config.gitlab_oauth_client_id} in gitlab oauth applications...")
+        logging.info(
+            f"Setting up client id {config.gitlab_oauth_client_id} and "
+            f"callback uris {gitlab_callback_uris} in gitlab oauth applications..."
+        )
+        curs.execute(
+            gitlab_oauth_cleanup,
+            {
+                "client_id": config.gitlab_oauth_client_id,
+                "redirect_uris": gitlab_callback_uris,
+                "client_secret": config.gitlab_oauth_client_secret,
+            },
+        )
```

Comment:
I will move this into a separate class. The dbInit class is general and what we use for all databases.

This is something extremely sketchy we do with the built-in gitlab database. We really should not be doing this in the first place. But since we barely use the built-in gitlab anyhow this rarely runs.

Reply:
Moved to a separate class.